Added to_spacy() functionality to convert dataframe of annotations to spacy trainable format (.spacy)

LeafmanZ opened the pull request:
I added to_spacy() functionality to convert dataframe of annotations to spacy trainable format (.spacy). Also added an extra line to the end of the demo notebook to show how the function is used.

LeafmanZ committed:
Update __init__.py

Added functionality to convert the dataframe of annotations into a spacy 3 trainable format (.spacy)

LeafmanZ committed:
Adding the use of the to_spacy() function to demo

Show how to use to_spacy() to save the dataframe of annotations into .spacy format for training in the Spacy3 pipeline.

LeafmanZ committed:
Merge branch 'ieriii:master' into master

philippnb commented:
Is there already a date planned for this feature to be released? :)

ieriii commented:
Hi! I'll review this PR as soon as possible. Thank for this and apologies for the delay.

philippnb commented:
Hi,
in the meantime I manually inserted the `to_spacy()` function in my Jupyter notebook.
And I was wondering whether it is possible to pause the annotation process? In my case I have a few hundreds of texts to label but want to split up the labeling in several days.
I tried exporting it as .csv file (`df_labels.to_csv('annotatedData.csv',index=False)`) and later on merging the csv files into one via
`merge = pd.concat([data1, data2])`
When I then call `to_spacy()` function again with the merged dataframe I always receive: `ValueError: too many values to unpack (expected 2)` Are you able to reproduce this with behavior?

Or in general: Is there a way to pause the annotation process or does labeling have to happen in one pass?

Thank you!

LeafmanZ commented:
Hmm. I was not able to reproduce this behavior. If you could send me what your data1 and data2 dataframes looked like I might be able to figure it out.

Another fix I think might work is for you to annotate what data you want in one sitting and save it as lets say:
annotator.to_spacy(df_1_labels, "spacy_labels_1.spacy")

Then the next time you label data you can save it as:
annotator.to_spacy(df_2_labels, "spacy_labels_2.spacy")

Then you can load back in the .spacy files with the following:
from spacy.tokens import DocBin

# load .spacy formats as docbins
spacy_l1 = DocBin().from_disk("spacy_labels_1.spacy")
spacy_l2 = DocBin().from_disk("spacy_labels_2.spacy")

# get the lengths of your amount of annotated examples
len_db_1 = len(spacy_l1)
len_db_2 = len(spacy_l2)

# combine your two annotated examples
spacy_l1.merge(spacy_l2)

# make sure the length of the two annotated examples equals the combined examples
assert(len_db_1 + len_db_2 == len(spacy_l1))

spacy_l1.to_disk('combined_spacy_labels.spacy')

Once this pull request is approved I may make another function that deals with merging multiple .spacy annotated data.

philippnb commented:
Thanks for your reply! I figured out that the csv files were read as string, but supposed to be tuples instead. 
This thread shows the same issue I experienced and how to fix it: https://stackoverflow.com/questions/61589488/convert-a-column-in-a-dataframe-from-type-string-to-tuple

Loading the .spacy files as DocBin and then merging them is also something I will try out!

ieriii commented on demo/spacy_annotator_demo.ipynb:
```diff
@@ -254,7 +254,7 @@
    "metadata": {},
    "source": [
-    "### Inspect output"
+    "### Inspect output and save dataframe of annotations to .spacy format for training in Spacy3 pipeline."
```
I would change it to:   
`## Inspect output and save dataframe of annotations to .spacy format for training in Spacy3 pipeline`

ieriii commented on demo/spacy_annotator_demo.ipynb:
```diff
@@ -264,7 +264,8 @@
    "source": [
-    "df_labels"
+    "df_labels\n",
+    "annotator.to_spacy(df_labels, \"spacy_labels.spacy\")"
```
I  would split line 267 and 268 in two separate cell blocks.
Also, I would change line 268 to:
```python
spacy_annotations = annotator.to_spacy(df_labels, "spacy_labels.spacy")
```

ieriii commented on spacy_annotator/__init__.py:
```diff
@@ -135,6 +138,50 @@ def __add_annotation(self, df, col_text, current_index, annotations):
         entities = {"entities": spans}
         df.at[current_index, "annotations"] = (df[col_text][current_index], entities)
+    
+    def to_spacy(self, df, file_path: str):
```
I would move this method _after_ the `annotate` method to keep the flow of the class from load data, to annotate to export in spacy format.

ieriii replied:
`filepath` requires a default.
I would put `filepath = None` as default and add logic to the method to export spacy format with a default name to current directory, for example:

```python
if file_path is None:
   file_path = os.path.join(os.getcwd(), 'annotations.spacy')
```

ieriii commented on spacy_annotator/__init__.py:
```diff
@@ -135,6 +138,50 @@ def __add_annotation(self, df, col_text, current_index, annotations):
+
+        # convert the docbin into a .spacy file for training in a spacy pipeline
+        db.to_disk(file_path)
+        print("Saved spacy file to:", os.path.join(os.getcwd(), file_path))
```
See comment above.   This should be changed to:   
```python
print(f"Spacy file saved to: {file_path}"))
```

ieriii commented on spacy_annotator/__init__.py:
```diff
@@ -135,6 +138,50 @@ def __add_annotation(self, df, col_text, current_index, annotations):
+        Parameters
+        ----------
+        df (pandas dataframe): Dataframe returned by the annotator (see Annotate()).
+        file_path (string): Filepath to save the .spacy file to.
```
file_path (string): Filepath (including filename) to save the .spacy file to.

ieriii commented:
@LeafmanZ, many many thanks for your PR and apolgoies for the late reply.
I have reviewed and requirested some changes. Grateful if you can address them. I'm happy to merge after that.

Let me know if anything is unclear or need further guidance.   

Thank you again for the time you spent on this. Appreciated.   
ps: apologies if some comments looks pedantic.

LeafmanZ committed:
Add files via upload

LeafmanZ committed:
Delete spacy_labels.spacy

LeafmanZ committed:
Add files via upload

LeafmanZ committed:
Add files via upload

LeafmanZ committed:
Add files via upload

LeafmanZ committed:
Add files via upload

LeafmanZ committed:
Add files via upload

LeafmanZ committed:
Add files via upload

LeafmanZ committed:
Delete __init__.cpython-310.pyc

LeafmanZ commented:
@ieriii Okay I think I fixed everything I think. I added a few more examples in the demo of ways to save the spacy files. Apologies ahead for the list of commits that you see above. I kept checking realized I had small changes to make after committing.

ieriii reviewed (approved):
LGTM.

I'll merge, make a few tweak to format and update package on pypi.
To do (it can be separate PR): alert the user in case filter span remove any overlapping spans.
